optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java

```diff
@@ -1966,6 +1968,121 @@ public void remove(Key key, Value value) {
                                 ConstraintCollectors::throwOnKeyConflict, TreeMap::new)));
     }
 
+    // Consecutive
+
+    /**
+     * Creates a constraint collector that returns {@link ConsecutiveData} about the first
+     * fact.
+     *
+     * For instance ${@code [Shift slot=1] [Shift slot=2] [Shift slot=4] [Shift slot=6]}
+     * returns the following infomation
+     *
+     * ${@code
+     * Consecutive Lengths: 2 1 1
+     * Break Lengths: 1 2
+     * Consecutive Items [[Shift slot=1] [Shift slot=2]], [[Shift slot=4]] [[Shift slot=6]]
+     * }
+     * 
+     * @param indexMap Maps the fact to its position in the sequence
+     * @param <A> type of the first mapped fact
+     * @return never null
+     */
+    public static <A> UniConstraintCollector<A, ConsecutiveSetTree<A, Integer, Integer>, ConsecutiveData<A, Integer>>
+            consecutive(ToIntFunction<A> indexMap) {
```

Comment:
We need to figure out a way to support `LocalDateTime` (and similar types) out of the box. A shift has a startTime and and endTime, and we need to be able to process that "natively" into a sequence.